rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,460 @@
+
+We wall call this approach the **PARM** (**P**ackage **A**s **R**records
+**M**odel) thereafter.
+
+#### Specifying dependencies
+
+The `pkgs.hello` example unveils a difficulty about specifying dependencies. It
+would work in a model where all the packages are defined in one place and
+populate the same huge record that will give Nickelpkgs. But that would be
+totally impracticable and anti-modular.
+
+Writing the `pkgs.hello` in its own file in Nickel today would result in the
+`gtk` occurrence raising an `unbound variable` error, and for a good reason:
+`gtk` is used but seemingly defined nowhere. In practice, `gtk` is expected to
+appear in the final recursive record that would Nickelpkgs. But from a language
+point of view, allowing references to yet-to-be-defined identifiers, that is
+having dynamic scoping, comes with its share of drawbacks: non-locality making
+code harder to understand, inability to detect unbound variables statically
+(incurring late error reporting), inability to properly typecheck such code,
+etc.
+
+Let us review possible suctions:
+
+##### Explicit list
+
+One possibility is to specify explicitly all the
+dependencies as record fields without definition:
+
+```nickel
+# file hello.ncl
+{
+  # dependencies
+  gtk | NickelPackage,
+
+  name = "hello",
+  version = "1.12",
+  description = "A program that produces a familiar, friendly greeting",
+  license = licenses.gpl,
+
+  enable_gui
+    | doc "Enable GTK+ support."
+    | Bool
+    | default false,
+
+  src = builders.fetchurl & { url = ..., sha256 = ... },
+
+  buildInputs = if enable_gui then [ gtk ] else [],
+
+}
+```
+
+The contract may be optional as the infrastructure of an hypothetical Nickelpkg
+would already apply the contract Package.
+
+<!-- In this design, it would be nice to avoid having to repeat those
+     dependencies in several places (like in inputs). -->
+
+##### Dynamic scoping
+
+<!-- DRAFT. not sure this is a good idea -->
+
+A second solution would be to implement a form of dynamic scoping, but with a
+special syntax, such that it doesn't impact normal uses of variables. Could
+either have a special `self` or `super` keyword that refers to the final version
+of the fixpoint, or a special syntax such as identifiers starting a specific
+character:
+
+```nickel
+{
+  name = "hello",
+  version = "1.12",
+  description = "A program that produces a familiar, friendly greeting",
+  license = licenses.gpl,
+
+  enable_gui
+    | doc "Enable GTK+ support."
+    | Bool
+    | default false,
+
+  src = builders.fetchurl & { url = ..., sha256 = ... },
+
+  buildInputs = if enable_gui then [ super.gtk ] else [],
+
+}
+```
+
+**TODO**: other solutions? The pkg subfield, but seems like a lesser version of
+the first proposal. Bazel-like includes? From experience, the semantics is
+confusing, it's tied to the filesystem structure. And in the end it's still a
+strange form of dynamic scoping: if we are to do it, let go for a proper version.
```

Comment:
I agree with the confusing-ness of bazel targets, **but** they have a very strong point for them in that because since they are at a meta-language level, they are only resolved at the very last moment, meaning that
1. The actual path of the dependency really stays fully opaque, which
    1. Is more robust. For example, there was a couple of occurrences of code manipulating the output path of a dependency in nixpkgs, and that broke with CA derivations so I had to change it − like https://github.com/NixOS/nixpkgs/pull/124227
    2. (potentially) makes the final UX nicer because you don’t have ugly store paths leaking everywhere whenever you want to do something
2. Evaluating a target doesn’t require evaluating its dependencies (meaning that the whole discussion about packages-as-data vs packages-as-functions doesn’t even have to exist)

We anyways couldn’t do something like what bazel does (because in the Nix model you do need to resolve the paths of your dependencies to generate the build script, which isn’t the case in bazel), but maybe an in-between solution could be interesting